rllib/policy/sample_batch.py

```diff
@@ -133,27 +134,68 @@ def __init__(self, *args, **kwargs):
         if self._is_training is None:
             self._is_training = self.pop("is_training", False)
 
+        # Try to infer the "length" of the SampleBatch by finding the first
+        # value that is actually a ndarray/tensor.
         lengths = []
         copy_ = {k: v for k, v in self.items() if k != SampleBatch.SEQ_LENS}
         for k, v in copy_.items():
             assert isinstance(k, str), self
 
-            # TODO: Drop support for lists as values.
+            # TODO: Drop support for lists and Numbers as values.
             # Convert lists of int|float into numpy arrays make sure all data
             # has same length.
-            if isinstance(v, list):
+            if isinstance(v, (Number, list)):
                 self[k] = np.array(v)
 
-            # Try to infer the "length" of the SampleBatch by finding the first
-            # value that is actually a ndarray/tensor. This would fail if
-            # all values are nested dicts/tuples of more complex underlying
-            # structures.
+            if k == SampleBatch.INFOS:
+                # Don't attempt to count on infos since we make no assumptions
+                # about its content
+                continue
+
+            # If this is a nested dict (for example a nested observation),
+            # try to flatten it, assert that all elements have the same length (batch
+            # dimension)
+            v_list = tree.flatten(v) if isinstance(v, (dict, tuple)) else [v]
+            # TODO: Drop support for lists and Numbers as values.
+            # If v_list contains lists or Numbers, convert them to arrays, too.
+            v_list = [
+                np.array(_v) if isinstance(_v, (Number, list)) else _v for _v in v_list
+            ]
             try:
-                len_ = len(v) if not isinstance(v, (dict, tuple)) else None
+                # Add one of the elements' length, since they are all the same
+                len_ = len(v_list[0])
                 if len_:
                     lengths.append(len_)
```

Comment:
just to make sure I understand this, we queue this ``len_`` only if ``len(xxx)`` does not throw error.
so it's possible to mix up things that have lengths and not have lengths?
asking because I see that you want to get rid of the ``None`` support below.

Reply:
Scalar np arrays don't have a length!

Reply:
Yep. Nones obviously don't have a length, but we are dropping them as you said.
But unsized arrays also don't have a length.
We currently support these as per our unit test. For example `SampleBatch({"obs": {"a":1, "b": 2}})` would fail otherwise.

Reply:
Or `SampleBatch({"obs": {"a":np.array(1), "b": np.array(2)}})` for that matter

Reply:
oh, that i scalar np array. I see.